### ISSUE TITLE
Reserve Support

### ISSUE THREAD

**Issue by creatorfromhell** [Rocologo/BagOfGold#67]
I am the developer of the plugin Reserve, a modern economy API with modern economy features in mind. As I am now ready to publicly push the API, I'm currently looking for developers to add support for it.

Link: https://github.com/TheNewEconomy/Reserve

Advantages over Vault:
- Multi-currency support
- Async support using Java's completeable futures
- Optional Extended API to allow third party plugins to add new currencies

Current Plugins that support Reserve:
- Towny
- ChestShop3
- MineBay

Current Providers/Planned Providers:
- The New Economy
- Towny Eco
- GemsEconomy

Thanks in advance,
creatorfromhell

**Comment by Rocologo** [comment]
Hi, BagOfGold does currently support Vault, but is it possible to support both Vault and Reserve? And how will they be able to work together? What happens when both Vault and Reserve is installed?

Im not sure if I understand the difference between the Async methods and the non-async methods?

What is a non-player? Give me an example...

**Comment by creatorfromhell** [comment]
> Hi, BagOfGold does currently support Vault, but is it possible to support both Vault and Reserve? And how will they be able to work together? What happens when both Vault and Reserve is installed?
> 
> Im not sure if I understand the difference between the Async methods and the non-async methods?
> 
> What is a non-player? Give me an example...

Vault and Reserve are able to work together. Of course, this is mostly based on how third-party plugins implement support for them, but if you want a good example I actually did the Reserve implementation into Towny, which supports both. In essence, this is how it works; There's a base class with various methods for getting balance, checking account existence, removing funds, etc. If there's a vault provider, then we can use that, but if there's a Reserve provider we use that as a priority due to the advantages over the base Vault implementation.

Async vs sync is very important for Economy Providers. For example, if an economy provider requires various calculations for item-based currencies, and you don't want to block the main server thread, you may utilize Reserve's async methods to utilize the result without blocking the main thread. The basic example is for third-party plugin commands that require a player's balance but may take the economy provider some time to calculate.

Non-player vs Player:
- Non-player could be a bank account, town in towny, faction in factions, guild, clan, server account, etc. This is to allow economy plugins to handle non-server accounts however they like, verse dealing with a UUID that is not tied to a player.

Player account
- This is fairly obvious, it's a player.

**Comment by Rocologo** [comment]
FYI. I have two plugins. bagOfGold, which is an item based economy like Gringotts, and then I have MobHunting where the players get rewards Money and heads and more for killing mobs. If I decide to support Reserve the I will add support in both plugins. But I am not ready to drop Vault because of the limited number of plugins which support Reserve so far. Vault and reserve will have to work together some how.

In the extended api you have something you call Tier. I dont understand what an economy tier is as well?

**Comment by creatorfromhell** [comment]
> FYI. I have two plugins. bagOfGold, which is an item based economy like Gringotts, and then I have MobHunting where the players get rewards Money and heads and more for killing mobs. If I decide to support Reserve the I will add support in both plugins. But I am not ready to drop Vault because of the limited number of plugins which support Reserve so far. Vault and reserve will have to work together some how.
> 
> In the extended api you have something you call Tier. I dont understand what an economy tier is as well? Is it Dollors and Cent?

As far as vault & Reserve working together from an economy provider POV(bagOfGold), Reserve automatically implements a default Vault provider class to support plugins that don't directly have reserve support. Of course, you can still include your own vault class, and they will work together.

A tier is a denomination. I.e. Penny, Quarter, Dollar, Ten Dollars in the US.

**Comment by Rocologo** [comment]
Ahh. I have actually thought about adding denominations like 10000 Coins = 100 BagOfGold = 1 Chest 

Okay I will see if I am able to add support for both plugins (Vault + Reserve) I don't think server owners will drop Vault because it is supported by so many other plugins. But I see the need for a new API with more features. So I will probably look into this.

**Comment by creatorfromhell** [comment]
Cheers, if you have any questions, or suggestions dont hesitate to contact me, I have a discord too(creatorfromhell#0666).

**Comment by Rocologo** [comment]
I finally found some time to look at this and have made some basic stuff.

Do you have any examples on how to an economy plugin hooks into 

1) Vautl

2) If Vault not present, the try Reserve

?

**Comment by Rocologo** [comment]
I have tried this, but I dont understand how to get reserveEconomy

	private void setupReserve() {
		reserveEconomy = ????
		Reserve.instance().registerProvider(reserveEconomy);
		getLogger().info("[BAGOFGOLD] Hooked into Reserve");
	}

**Comment by creatorfromhell** [comment]
There's Towny: https://github.com/TownyAdvanced/Towny

Also my EssentialsX PR: https://github.com/EssentialsX/Essentials/pull/2601

**Comment by Rocologo** [comment]
I didnt understand how to get reserveEconomy ?

**Comment by creatorfromhell** [comment]
From the wiki https://github.com/TheNewEconomy/Reserve/wiki/Economy-API

if(Reserve.instance().economyProvided()) {
  EconomyAPI api = Reserve.instance().economy().get();
}

**Comment by Rocologo** [comment]
Its an Economy Provider I am making???

I tried to use 

		if(Reserve.instance().economyProvided()) {
			reserveEconomy = Reserve.instance().economy().get();
		}

But ".get();" is not found when I "import net.tnemc.core.Reserve;"

When I look at the methods presented to me using "Reserve.instance().economy().*" then it looks like methods to be used by a plugin which hooks into reserve.. beginning with addHoldings(...).... This does not look right to me?

**Comment by Rocologo** [comment]
![image](https://user-images.githubusercontent.com/1008133/61148405-ff093800-a4de-11e9-8370-8b0ad3707644.png)

**Comment by Rocologo** [comment]
This is the guide I have been using, but it doesnt show how to get the reserveEconomy variable???
https://github.com/TheNewEconomy/Reserve/wiki/Economy-Provider

**Comment by creatorfromhell** [comment]
Ah my bad, I thought you were wanting to call the API, the reserveEconomy variable is the instance of your plugin's implementation of the EconomyAPI class, seen here: https://github.com/TheNewEconomy/Reserve/blob/master/src/net/tnemc/core/economy/EconomyAPI.java . If you're interested in the ExtendedAPI that is found here, it adds more strict rules such as specific currency class interfaces, etc https://github.com/TheNewEconomy/Reserve/blob/master/src/net/tnemc/core/economy/ExtendedEconomyAPI.java

**Comment by Rocologo** [comment]
Today I succeded implementing basic Reserve methods to my provider plugin BagOfGold. So I started to look closer at your EconomyAPI and implement some of the Methods. Im still a bit confused about the new names in your EconomyAPI so I have a few questions.

Vaults API has the following methods and I am not sure what they are called in EconomyAPI ?

Economy.isBankowner(String account)
Economy.bankBalance(String account)

**Comment by creatorfromhell** [comment]
Reserve currently doesn't have bank methods, mostly due to the fact I'm still sorting out the methods that make the most sense, and how much control I want Reserve to have over banks.

**Comment by Rocologo** [comment]
OH NO....In that case I cant implement Reserve... then you are not 100% compatible with Vault :-(

BagOfGold has Bank features and these are VERY important because this is the only way to store the bags safely.

I need minimum the following methods

hasBankSupport()
getBankBalance(Player player)
getBankBalance(Player player,String world)
isBankOwner(Player)
isBankMember(Player) (I dont use this at the moment but I might need it in the future)
bankDeposit(Player,amount)
bankWithdraw(Player,amount)
bankHas(player,amount)

**Comment by Rocologo** [comment]
I the moment I only use basic bank functions.

IE. Each player has only **1** balance and **1** bankBalance for each worldGroup in each GameMode. A worldGroup is one or more World. 

https://github.com/Rocologo/BagOfGold/blob/master/src/one/lindegaard/BagOfGold/PlayerBalance.java

A player cant have more than one bank account per worldGroup per Gamemode, and there is only one Currency "BagOfGold" at the moment, because Vault does not support Currencies.

I think you will need some minimum permissions for banks. But have tried speaking with one of the Bank plugin developers? What they need?


Im not able to add support for Reserve if there is no Bank support in Reserve.

**Comment by creatorfromhell** [comment]
Yeah, I've been in contact with some bank plugin creators, and am waiting on feedback.

**Comment by Rocologo** [comment]
I'm sorry but I have to stop implement Reserve until the interface has bank methods as well. Reserve needs to be 100% backwards compatible with vault if you want success with this project.

**Comment by creatorfromhell** [comment]
Here's what I came up with, let me know if you notice any methods missing. https://paste.ubuntu.com/p/zgVQBTbTtb/

**Comment by Rocologo** [comment]
I looks good. Lets go with that.

Im still not sure what Im going to do with all the Async**** Methods. Most of my operations cant be made async because the money is related to items/itemstacks inside the game in the players inventory. So at the moment I will let them return null/false.   

Im going on vacation today, but I will be online from time to time. Let me know when then changes is pushed to Github.

**Comment by Rocologo** [comment]
I will have to spend 4 hours alone in the airport/airplane so it would be nice to have then new methods So the sooner you can push it to Github, the better.

**Comment by creatorfromhell** [comment]
> I will have to spend 4 hours alone in the airport/airplane so it would be nice to have then new methods So the sooner you can push it to Github, the better.

The methods have been published.

**Comment by Rocologo** [comment]
I am currently testing. I have added support for both BagOfGold + MobHunting. I still have some bugs to find, but all need methods is implemented :-)

**Comment by Rocologo** [comment]
FINALLY. I have added support for Reserve in both my plugins

https://www.spigotmc.org/resources/mobhunting.3582/
https://www.spigotmc.org/resources/bagofgold.49332/

I have tested my Provider with ChestShop and of cause with MobHunting. 

I hope that you will add my plugins to the list of Implementations
https://github.com/TheNewEconomy/Reserve/wiki/Implementations
And on other lists you have. 

Thank you for being patient with me and help me understanding how to implement Reserve.

I have still all the Async methods to do, I am not sure what to do with them.
